Update exported symbols and add ABI check

mscastanho opened the pull request:
libnxz is currently exporting many symbols that are of internal use, mainly global variables. With the code in this PR all such symbols have been made internal, either by declaring them static, or moving into `nx_config` or even turning into `#define`s. There are still a few cases in which I'm not sure if any of the previous options apply, so I explicitly added them to the list of local symbols in the linker script.

Now we're down from 22 exported variables to 0.

I also included a test using abidiff to check ABI compatibility with zlib and to guarantee we spot ABI changes when they happen. A good way to test it is to remove one function from the list of `missing_symbols` in `test/test_abi` and run it again. Also, try to remove the `static` keyword from, for example, `timeout_seconds` in `lib/gzip_vas.c`. Both of these should demonstrate how the test will catch ABI changes.
 
There are some other misc changes as well, but I tried to separate each in their own commit to facilitate reviewing.

Closes #25

tuliom commented:
In the 7th commit you added:
> Make all nx_* symbols private

What does that mean? There are many `nx_*` symbols that are still global and they should continue global.

tuliom commented on test/test_abi:
```diff
@@ -0,0 +1,94 @@
+trap "rm -f ${supps_file} ${zlib_abi}" EXIT
+
+# Generate suppressions file
+echo > ${supps_file}
```
Any reason for generating this dynamically, instead of maintaining it in the repository?

mscastanho replied:
It was an attempt to make it more self-contained. I'm OK with putting it in a separate file as well

tuliom commented on test/test_abi:
```diff
@@ -0,0 +1,94 @@
+
+# These should be either the .so files or their descriptions generated by abidw
+zlib_abi_orig="libz.1.2.11.abi"
+libnxz_abi="libnxz.0.63.0.abi"
```
It will be very hard to maintain these tests as zlib and libnxz have newer releases.
Can we use `zlib.abi` and `libnxz.abi` instead?

mscastanho replied:
Sure, I'll change that in the next push.

tuliom commented on test/libnxz.0.63.0.abi:
```diff
@@ -0,0 +1,2548 @@
+    <elf-symbol name='uncompress2' version='ZLIB_1.2.9' is-default-version='yes' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='zlibVersion' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+  </elf-function-symbols>
+  <abi-instr version='1.0' address-size='64' path='crc32_ppc.c' comp-dir-path='/home/mscastanho/src/power-gzip/lib' language='LANG_C11'>
```
What are these paths doing here?
Can we get rid of them?

mscastanho replied:
Done.

tuliom commented on test/libnxz.0.63.0.abi:
```diff
@@ -0,0 +1,2548 @@
+    <typedef-decl name='__useconds_t' type-id='type-id-1' filepath='/usr/include/bits/types.h' line='161' column='1' id='type-id-5'/>
+    <var-decl name='nx_fault_storage_address' type-id='type-id-6' visibility='default' filepath='/home/mscastanho/src/power-gzip/lib/nx_zlib.h' line='521' column='1'/>
+    <var-decl name='tb_freq' type-id='type-id-7' visibility='default' filepath='../inc_nx/nxu.h' line='71' column='1'/>
+    <function-decl name='usleep' filepath='/usr/include/unistd.h' line='469' column='1' visibility='default' binding='global' size-in-bits='64'>
```
This is a system function in a system file.
Can we avoid tracking this?

mscastanho replied:
Done.

tuliom reviewed (changes requested):
The code looks very good. I have a few comments and suggestions.

I missed:
1. A configure check for `abidiff` that fails if it isn't available in the system.
2. A way (a make target?) to regenerate both abi files.

mscastanho commented:
> What does that mean? There are many nx_* symbols that are still global and they should continue global.

@tuliom That's bad wording on my part. I meant to say to grup all local `nx_*` into a single wildcard statement in the linker script, instead of one for each. All global `nx_*` symbols are explicitly listed and will continue to be global. I'll update the commit message.

mscastanho commented:
@tuliom I believe the latest push addresses all your comments. I just couldn't find a simple way to get the path to the system's libz.so to help regenerating `libz.abi` automatically, but I added instructions in the form of message to the user.

tuliom commented on test/run_test.sh:
```diff
@@ -8,57 +8,30 @@ run_test_report=run_test_report_${TS}.txt
+    fi
+}
+
+run ./test_abi ../lib/libnxz.so
```
Is this passing even when using `--drop-undefined-syms`?

mscastanho replied:
Not yet, I'm still trying to understand what is going on there. If we stop using `--drop-undefined-syms` when generating the `.abi` files, then the issue disappears.

mscastanho replied:
@tuliom Those were being caused by some incompatible function declarations for `crc32` and `nx_adler32_combine`. I added a new commit to address that, regenerated `libnxz.api`, and now all tests pass.

tuliom reviewed (commented):
The code LGTM.
I have just one question.

mscastanho commented on configure.ac:
```diff
@@ -21,6 +21,10 @@ AC_CANONICAL_HOST
+AC_CHECK_PROG(ABIDIFF, abidiff, abidiff, no)
+AC_CHECK_PROG(ABIDW, abidw, abidw, no)
+AS_IF([test "x$ABIDIFF" = xno -o "x$ABIDW" = xno],
+      [AC_MSG_ERROR([Please install abidiff and abidw (tools from libabigail).])])
```
I just removed the second one in the latest push.

tuliom commented on configure.ac:
```diff
@@ -21,6 +21,9 @@ AC_CANONICAL_HOST
 # Checks for programs.
+AC_CHECK_PROG(ABIDIFF, abidiff, abidiff, no)
+AS_IF([test "x$ABIDIFF" = xno],
+      [AC_MSG_ERROR([Please install abidiff (tool from libabigail).])])
```
Are there any issues preventing the usage of `AX_WITH_PROG(ABIDIFF, abidiff)`?

mscastanho replied:
I just didn't see the need for the extra `--with-$TOOL` option that comes with it. But I can add that to the PR.

mscastanho replied:
Also, just using `AX_WITH_PROG(ABIDIFF, abidiff)` doesn't make the build fail at configure time if `abidiff` is not available.

tuliom replied:
Ack.  Got it.
I'm resolving this conversation.

tuliom commented on lib/nx_crc.c:
```diff
@@ -466,7 +466,7 @@ uLong nx_crc32_combine64(crc1, crc2, len2)
     unsigned long crc;
     const unsigned char FAR *buf;
-    uint64_t len;
+    uInt len;
```
This is wrong in a 64-bit architecture because `uInt` becomes an `unsigned int`, which is 32-bit.

mscastanho replied:
@tuliom but `unsigned int` is exactly what is used by zlib. If the user needs something larger than 32 bits, it should be using `crc32_z` instead.

tuliom replied:
Ack.  I'm resolving this conversation.
Thanks!

tuliom commented on lib/nx_zlib.h:
```diff
@@ -525,7 +525,7 @@ extern uint64_t nx_wait_ticks(uint64_t ticks, uint64_t accumulated_ticks, int do
 /* zlib crc32.c and adler32.c */
 extern unsigned long nx_crc32_combine(unsigned long crc1, unsigned long crc2, uint64_t len2);
-extern unsigned long nx_adler32_combine(unsigned long adler1, unsigned long adler2, uint64_t len2);
+extern unsigned long nx_adler32_combine(unsigned long adler1, unsigned long adler2, off_t len2);
```
You can't change this declaration without changing `lib/nx_adler_combine.c` too.
The function is still treating len2 as `uint64_t`.

tuliom commented on libnxz.h:
```diff
@@ -154,11 +154,11 @@ extern int inflateSetDictionary(void *strm, const unsigned char *dictionary,
 extern ulong adler32(ulong adler, const char *buf, uint len);
 extern ulong adler32_z(ulong adler, const char *buf, size_t len);
-extern ulong crc32(ulong crc, const unsigned char *buf, uint64_t len);
+extern ulong crc32(ulong crc, const unsigned char *buf, unsigned int len);
```
Likewise here.

tuliom replied:
I think you meant to use `uInt` here.

mscastanho replied:
Ok, I'll the typedef.

tuliom reviewed (changes requested):
@mscastanho I'm afraid the new patch is broken and I didn't understand the changes to `configure.ac`.

mscastanho commented:
@tuliom I added the latest changes we discussed and also rebased on top of develop.
